DSL for PlasticView Anchors and Values

madbat opened the pull request:
This PR adds a small DSL intended to improve readability and speed of development when laying out `PlasticView`s, i.e. setting `Anchor`s with offsets.

The current way to set an anchor to a `PlasticView` with an offset is something like:
```swift
button.setTop(view.top, offset: .scalable(300))
button.setRight(view.right, offset: .fixed(-50))
```

With this DSL, the above example would turn into:
```swift
button.top = view.top + 300
button.right = view.right - 50.fixed
```

The contents of this PR are (see each commit message for more details):

- Add `offset` property to `Anchor` 

- Add ability to add an offset to an anchor through `+` and `-` operators

- Make `Value` conform to `ExpressibleByIntegerLiteral` and `ExpressibleByFloatLiteral`

- Add `scaled` and `fixed` properties to `CGFloat`, `Double` and `Int` as a convenient way to instantiate `Value` instances

bolismauro commented:
I'll review it this evening.
Some feedback
- remove `.scalable` and `'fixed` 
- Remove the `init(value)` (also from Size and EdgeInsets)
- Make sure all the examples work
- Check the documentation / readme and make sure that everything is updated

bolismauro commented:
I have a huge doubt about this.
We are introducing `30.fixed` and because of this, we should remove `.fixed(30)` (otherwise we have multiple methods to do the same thing, which is not good.
What about size and edgeInsets? We can't use `.fixed` anymore, for consistency. How would you replace them?

Another thing I don't like is that you changed the `Anchor` struct by adding the offset. An `Anchor` represents an edge of a view, and adding an offset doesn't really make sense...
Is it possible to implement the functionalities you want without doing this?

Third and last one, can you make a complete list of things you are adding? I saw several operator overloads and I'd like to have a list (e.g, to create the changelog of the bump) of things added/removed

madbat commented:
@bolismauro thanks for your thoughts on the PR!

Let me address them:

> We are introducing 30.fixed and because of this, we should remove .fixed(30) (otherwise we have multiple methods to do the same thing, which is not good.

The fact is: I'm not sure that I agree with this.

I **do agree** that having to renounce to the ability to have a nice way to instantiate sizes and edge insets – i.e. by using `.scalable(...)` or `.fixed(...)` – would be bad.

In fact, I wouldn't remove them! I see the extensions to `Int` and `Double` that I proposed in this PR as a nice, harmless, convenience, not intended to replace anything. It just felt like a really spot-on use case to be able to get a `Value` (which I see as an implementation detail I'd rather not be concerned about) from an Int or a Double.

As for your concern regarding multiple ways to do the same thing: I'm not convinced it's intrinsically bad, as long as we're talking about things that differ just _slightly in syntax_ and do not have any other bigger implication that one should be aware of.

So, I'd rather give people a bit more flexibility to adapt to different needs/contexts, and leave to each project's guidelines to set the rules for this kind of details.

That said, no big deal if people feel we're better off without them :)

> Another thing I don't like is that you changed the Anchor struct by adding the offset. An Anchor represents an edge of a view, and adding an offset doesn't really make sense... Is it possible to implement the functionalities you want without doing this?

I think it's the only solution to achieve something like `button.top = view.top + 300`.

I kind of agree that `offset` doesn't _really_ belong to an `Anchor` if you think of it as "an edge of a view".

**But**, if you think of an `Anchor` as "a mechanism to link one side of a view to a side of another view, with a — possibly non zero — offset between them", it suddenly makes more sense.

> Third and last one, can you make a complete list of things you are adding? I saw several operator overloads and I'd like to have a list (e.g, to create the changelog of the bump) of things added/removed

Sure thing!

sroddy commented:
If we are going to add the offset property directly to Anchors, I suggest to follow one single approach and entirely remove all the set*(anchor:, offset:) and the other convenience methods that use the offsets and just keep the top/bottom/left/right properties.
This will avoid the unnecessary confusion of having multiple offsets within a single set*() operation.

Neither I am 100% convinced but for sure this approach would:
- Simplify the interface removing a lot of convenience methods.
- Solve one of the biggest problems i had so far with the good old Plastic, which was having an offset in the views.pl_centerBetweenTop:andBottom: (and also seems to be an issue even in the current version of Plastic).

Another thing that I've noticed is that this new Anchor's `offset` property currently has no impact on the `coordinate` getter. Is that on purpose? And if so, don't you think this might be even more misleading?

madbat commented:
> Another thing that I've noticed is that this new Anchor's offset property currently has no impact on the coordinate getter. Is that on purpose? And if so, don't you think this might be even more misleading?

It's not on purpose. Thanks for making me notice it. Adding `offset` to the `coordinate` would be the right thing to do if we decide to go with my proposed changes.

bolismauro commented:
@sroddy 
```
Solve one of the biggest problems i had so far with the good old Plastic, which was having an offset in the view.pl_centerBetweenTop:andBottom: (and also seems to be an issue even in the current version of Plastic).
```
can you elaborate?

sroddy commented:
Here is a concrete example I faced a couple of months ago. Suppose you have to layout something like this:
![layout example 001](https://cloud.githubusercontent.com/assets/729934/20554737/23ab67c2-b15e-11e6-8d9c-8e15c8e94b9f.jpeg)

The red horizontal line is meant to be placed in the middle of the space between the bottom of the Title and the bottom of the View plus an offset that is highlighted with the red arrow.

In the current version of Plastic this layout is a bit tricky to achieve and most of the solutions involve the creation of transparent views (e.g. a transparent line placed below the title anchor with an offset) which serve no other purpose than aiding in the layout.

With the proposed change, this would be easily achieved like that: view.centerBetween(top: title.bottom + 20.scaled, bottom: root.bottom).

I guess this won't be the only example where adding an offset to anchors would simplify things.

bolismauro commented:
There are ways to do it without additional views but I get your point. Ok so, let's add this concept of offset for the anchors.

I think we should solve the issue with size/edgeinsets and then we can proceed with the implementation of everything.
I still don't have a clear idea about how to solve it. To me, we should try to be consistent though

madbat commented:
I've updated my PR to:

- add `offset` to `Anchor`'s `coordinate`

- remove `setTop/Bottom/Left/Right/CenterX/CenterY/Width/Height` methods from PlasticView, as they're superseded by `top/bottom/left/right/centerX/centerY` setters

- update `PlasticView` convenience methods – and relative tests –  that depend upon the aforementioned methods.

- Remove `Value` initializer `init(_ scalable: CGFloat)`

As for the debated `Int`, `Double` and `CGFloat` extensions that produce `scalable` and `fixed` values, I can remove them from this PR, and we may discuss about their appropriateness in another PR.

bolismauro commented on Katana/Plastic/Anchor.swift:
```diff
@@ -51,27 +56,32 @@ public struct Anchor: Equatable {
       
     }
+
+    coord += view.scaleValue(offset)
```
I'd change this in `return coord + view.scaleValue(offset)`

madbat replied:
👍

bolismauro commented on Katana/Plastic/Value.swift:
```diff
@@ -145,3 +187,65 @@ public struct Value: Equatable {
+ This allows `Value` instances to be initialized by integer and floating point literals,
+ having the scalable part equal to the numeric literal, and the fixed part equal to 0.
+ */
+extension Value: ExpressibleByIntegerLiteral, ExpressibleByFloatLiteral {
```
Can we remove this?

We can define the `+` operator also for Anchor/Int and Anchor/Double.
I don't really like these protocols to be honest
@smaramba @lucaquerella  opinions?

madbat replied:
In my opinion it's better to have `Value` conforming to `ExpressibleByIntegerLiteral` and `ExpressibleByFloatLiteral`, since they exist for this purpose: they make it clear to users of `Value` why using integer/float literals in operations expecting `Value`s work.

Defining an overload of the `+` operator for Anchor/Int and Anchor/Double would be more confusing IMHO.

bolismauro replied:
`they make it clear to users of Value why using integer/float literals in operations expecting Values works.` to me, they don't make it clear. You have to discover that you have implemented that protocol somewhere to understand the code, otherwise you don't get why it works. 

If you define the proper + operator, if is oblivious the reason why it works. 

In general we decided to don't use them multiple times because they can create confusing situations..
For instance (not tested, but it should work)
```swift
init(width: Value, height: Value) -> Size

// you can do
Size(width: 5, height: 7)
```
You are basically propagating this ability to use double instead of values everywhere and to me this is not good

madbat replied:
Ok, this discussion would go the same way the discussion on `.scalable/.fixed` went 😄

Should I proceed and remove all these additions from the PR, or would you prefer to learn @lucaquerella's and @smaramba's opinions first?

bolismauro replied:
I asked opinions to both. As always I will be happy with the decision of the majority of us.

I really don't get why I'm the only once concerned about losing clarity. With that additions you can swap numbers/Values.

smaramba replied:
checking 👍

smaramba replied:
I would say that I prefer ExpressibleByIntegerLiteral over the overloading of the `+` operator.
Motivation: having a `+` operator between two heterogeneous types is more mysterious than saying that an `Int` is automatically also a valid `Value` type. BUT: the thing we should discuss to me is if we want that 200 is equivalent to `.scalable(200)` or `.fixed(200)`

bolismauro replied:
`Motivation: having a + operator between two heterogeneous types is more mysterious than..` but then why it is ok?
https://github.com/BendingSpoons/katana-swift/pull/76/files#diff-1c9939d1c3f55e132473f60fbc4c2739R100

I'm just trying to understand how to decide what is better and what is worse.

To me, already not being able to create a fully Value from an int is a signal that we shouldn't convert them. 

Plus try to think about a newcomers in katana/swift. It is better to explain that + is defined also for int and anchor or to say "you know.. numbers are automatically converted into value.. scalable [/fixed] value.. but not always. You need to look at the context and understand when it is converted"

Probably I'm not yet into the swift way of thinking, but the first explanation is way way more easy to understand.

PLUS: we are introducing this behaviour for a very specific case. Why should we do something that spread this behaviour everywhere when we can limit it (and limit the cases where this confusion may arise basically)
?

smaramba replied:
`but then why it is ok?` well, but in that case we were not deciding between two approaches, that was the most natural one. In the LiteralConvertible vs  + case we have two options, both makes sense, but, saying that 200 is a .fixed/scalable(200) is a natural extension to me. At the end of the day you can create a `Value` starting from an Int using Value.fixed(200), this means that a Value and an Int can express the same thing.
I'm not saying that is good to introduce that per se, and that cannot lead to confusion in certain cases but I'm saying that `Value` and `Int` are interchangeable concepts in this specific context.

bolismauro replied:
Alright guys.. as you wish..
It still doesn't make any sense to me to create confusion everywhere just to handle a single case. I'm sure that it will lead to confusion in the future but of course I can't prove it.

```
but, saying that 200 is a .fixed/scalable(200) is a natural extension to me. At the end of the day you can create a Value starting from an Int using Value.fixed(200), this means that a Value and an Int can express the same thing.
```
1- A value is composed by two int (or double..). We should take a decision whether a single int is the same as a Value with fixed = 0 or scalable = 0. And then explain this to everyone 

2 - `Value.fixes(Int)` to me is like `Int(AString)`. I don't think anyone would automatically convert strings into int, right? They are different things, formed by different type of values. For the sake of simplicity during the layout phase we decided to allow the usage of numbers. It is cleaner and simpler to read. Ok fair enough (consider also that, in 99% of the cases, that number will be a scalable.. and here we also have a reason to pick one of the two). But this is a DSL for that specific case. Extend this DSL to everything, where things may be more complex, hard to understand and use cases change.. it is a BIG mistake to me

smaramba replied:
I'm kinda neutral, you are strongly against, so it's a no go to me. 👍

bolismauro replied:
```But I also believe that, to the user new to Katana that has never used Plastic, reading code filled with .scalable(...), Value(scalable: ..., fixed: ...) would be quite puzzling. ```
You can't use a library without knowing the basic concepts. Value is a basic concept of Plastic. Because of this I don't buy the puzzling argument. I agree with the boilerplate and in fact **I agree with the layout DSL**. I just don't want to extend this DSL everywhere

```
"Plastic scales dimensions for you in order to make your views lay out consistently across different screen sizes. If you don't want something to be scaled, just use .fixed"
```
in plastic you have to use fixed also when things are scaled. I can give you infinite examples of this

smaramba replied:
as discussed offline with @bolismauro I do agree that going with the most conservative implementation (overloading of `+`) in this specific case should be the right option, avoiding to leak out the bridge Int<->Value to other parts of the framework/application

madbat replied:
👍 let's limit it to this specific use case as you propose then, it's probably the safest and more useful thing to do.

Besides adding the `+` and `-` operators with `Int` and `CGFloat`/`Double` versions, do you prefer we keep also the ones accepting `Value`s (in case I already have a `Value` instance I want to use as a offset – which would be rare) or we remove them?

bolismauro commented on Katana/Plastic/Value.swift:
```diff
@@ -145,3 +187,65 @@ public struct Value: Equatable {
+  }
+}
+
+extension CGFloat {
```
I'd remove these from this PR and move this discussion in another issue/PR

madbat commented:
> Can you also make sure that
> 
> - all the examples work
> 
> - the documentation is updated
> 
> - the readme is updated
> 
> - you have added relevant tests, if needed

Everything on the list should be good (no readme changes nor new tests needed).

> Plus (I can't do it myself since you forked the repo) I'd remove the following methods

Done.

madbat commented:
I've

- 🔥 the conformance of `Value` to `ExpressibleByIntegerLiteral` and `ExpressibleByFloatLiteral`

- 🔥 extensions to `CGFloat`, `Double` and `Float` that implemented the Value-returning methods `scalable` and `fixed`

- added overloads to `+` and `-` operators operating on `Anchor` and `CGFloat` values, with corresponding tests.
